core/any_error.rs

```diff
@@ -0,0 +1,55 @@
+use std::any::{Any, TypeId};
+use std::convert::From;
+use std::error::Error;
+use std::ops::Deref;
+
+pub type ErrBox = Box<dyn AnyError>;
+
+pub trait AnyError: Error + Any + Send + Sync + 'static {}
+impl<T> AnyError for T where T: Error + Any + Send + Sync + Sized + 'static {}
+
+impl<T: AnyError> From<T> for Box<dyn AnyError> {
+  fn from(error: T) -> Self {
+    Box::new(error)
+  }
+}
+
+pub trait DowncastBox: Sized {
+  fn downcast_box<T: Any>(self) -> Result<Box<T>, Self>;
+}
+
+impl<S> DowncastBox for S
+where
+  S: Sized + Deref + Into<Box<<S as Deref>::Target>>,
+  <S as Deref>::Target: Any,
+{
+  fn downcast_box<T: Any>(self) -> Result<Box<T>, Self> {
+    if Any::type_id(&*self) == TypeId::of::<T>() {
+      let target = Box::into_raw(self.into()) as *mut T;
+      let target = unsafe { Box::from_raw(target) };
+      Ok(target)
+    } else {
+      Err(self)
+    }
+  }
+}
+
+pub trait DowncastRef {
+  fn downcast_ref<T: Any>(&self) -> Option<&T>;
+}
+
+impl<S> DowncastRef for S
+where
+  S: Deref,
+  <S as Deref>::Target: Any,
+{
+  fn downcast_ref<T: Any>(&self) -> Option<&T> {
+    if Any::type_id(&**self) == TypeId::of::<T>() {
+      let target = self as *const Self as *const &T;
+      let target = unsafe { &(*target as &T) };
+      Some(target)
+    } else {
+      None
+    }
+  }
+}
```

Comment:
It would be nice to have a small unit test here for `downcast_ref` and `downcast_box` on mock error types.